File: data_structure/prefix_sum/560_subarray_sum_equals_k.cc
#include <iostream>
#include <numeric>
#include <unordered_map>
#include <vector>

// 此方法会超时.
int SubarraySum(std::vector<int>& nums, int k) {
  std::vector<int> sums(nums.size() + 1, 0);
  std::partial_sum(nums.begin(), nums.end(), sums.begin() + 1);
  int ans = 0;
  for (int i = 0; i < static_cast<int>(nums.size()); ++i) {
    for (int j = i; j < static_cast<int>(nums.size()); ++j) {
      if (sums[j + 1] - sums[i] == k) {
        ++ans;
      }
    }
  }
  return ans;
}

// 使用一个哈希表, key 是前缀和, value 是该前缀和出现的次数.
// 假设当前的前缀和是 prefix_sum, 那么 hash_map[prefix_sum - k] 即以当前位置结尾, 满足条件的区间个数.
int SubarraySum2(std::vector<int>& nums, int k) {
  int count = 0, prefix_sum = 0;
  std::unordered_map<int, int> hash_map;
  hash_map[0] = 1;
  for (int num : nums) {
    prefix_sum += num;
    count += hash_map[prefix_sum - k];
    ++hash_map[prefix_sum];
  }
  return count;
}

int main(int argc, char* argv[]) {
  std::vector<int> nums = {1, 2, 3};
  std::cout << SubarraySum2(nums, 3) << std::endl;
  return 0;
}

File: data_structure/prefix_sum/560_subarray_sum_equals_k.py
from collections import defaultdict
from typing import List

# 2026/01/18 [left, right), 此方法会超时
def subarray_sum(nums: List[int], k: int) -> int:
    sum_list = [0]
    acculate_sum = 0
    for num in nums:
        acculate_sum += num
        sum_list.append(acculate_sum)

    res = 0
    for left in range(len(nums)):
        for right in range(left + 1, len(nums) + 1):
            interval_sum = sum_list[right] - sum_list[left]
            if interval_sum == k:
                res += 1

    return res

# 2026/01/18 抄答案, 思路分析见 .cc 文件.
def subarray_sum2(nums: List[int], k: int) -> int:
    result = 0
    prefix_sum = 0
    hash_table = defaultdict(int)

    hash_table[0] = 1

    for num in nums:
        prefix_sum += num
        result += hash_table[prefix_sum - k]
        hash_table[prefix_sum] += 1

    return result

nums = [1,2,3]
k = 3
print(subarray_sum2(nums, k))
